File: src/hbse/core/redaction.py
"""Redaction primitives for controlled output and audit safety."""

from __future__ import annotations

import base64
import hmac
import re
from urllib.parse import quote
from hashlib import sha256

from hbse.core.serialization import b64url_no_padding


class RedactionLeakError(ValueError):
    """Raised when a controlled output still contains a known secret."""


class RedactionEngine:
    def __init__(self, fingerprint_key: bytes, fingerprints: list[str] | None = None) -> None:
        self._fingerprint_key = fingerprint_key
        self._fingerprints = set(fingerprints or [])
        self._known_values: dict[str, bytes] = {}

    def fingerprint(self, secret: bytes) -> str:
        digest = hmac.new(self._fingerprint_key, secret, sha256).digest()
        return b64url_no_padding(digest[:18])

    def learn(self, secret: bytes) -> str:
        fingerprint = self.fingerprint(secret)
        self._fingerprints.add(fingerprint)
        self._known_values[fingerprint] = secret
        return fingerprint

    def redact_text(self, text: str) -> str:
        redacted = text
        for fingerprint, secret in self._known_values.items():
            marker = f"[REDACTED:secret:{fingerprint}]"
            for representation in self._representations(secret):
                if representation:
                    redacted = redacted.replace(representation, marker)
        redacted = self._redact_structured_patterns(redacted)
        return redacted

    def assert_no_known_secret(self, text: str) -> None:
        redacted = self.redact_text(text)
        if redacted != text:
            raise RedactionLeakError("controlled output contains a known secret")

    @staticmethod
    def _representations(secret: bytes) -> set[str]:
        values: set[str] = set()
        try:
            secret_text = secret.decode("utf-8")
            values.add(secret_text)
            values.add(quote(secret_text, safe=""))
        except UnicodeDecodeError:
            pass
        values.add(base64.b64encode(secret).decode("ascii"))
        values.add(base64.urlsafe_b64encode(secret).decode("ascii").rstrip("="))
        return values

    @staticmethod
    def _redact_structured_patterns(text: str) -> str:
        patterns = [
            (re.compile(r"(?i)(Authorization:\s*Bearer\s+)[^\s,;]+"), r"\1[REDACTED:bearer]"),
            (re.compile(r'(?i)(password=)(["\']?)[^;&\s"\']+(\2)'), r"\1\2[REDACTED]\3"),
            (re.compile(r'(?i)(token=)(["\']?)[^;&\s"\']+(\2)'), r"\1\2[REDACTED]\3"),
            (re.compile(r'(?i)(api[_-]?key=)(["\']?)[^;&\s"\']+(\2)'), r"\1\2[REDACTED]\3"),
            (re.compile(r'(?i)("?(?:password|token|api[_-]?key)"?\s*:\s*")[^"]+(")'), r"\1[REDACTED]\2"),
        ]
        redacted = text
        for pattern, replacement in patterns:
            redacted = pattern.sub(replacement, redacted)
        return redacted

File: hbse/src/redaction.rs
use hmac::{Hmac, Mac};
use regex::Regex;
use sha2::Sha256;

use crate::serialization::b64url_no_padding;

type HmacSha256 = Hmac<Sha256>;

pub fn redaction_fingerprint(fingerprint_key: &[u8], secret: &[u8]) -> String {
    let mut mac = HmacSha256::new_from_slice(fingerprint_key).expect("HMAC accepts any key");
    mac.update(secret);
    let digest = mac.finalize().into_bytes();
    b64url_no_padding(&digest[..18])
}

#[derive(Debug, Clone)]
pub struct RedactionEngine {
    known_values: Vec<(String, Vec<u8>)>,
}

impl RedactionEngine {
    pub fn new() -> Self {
        Self {
            known_values: Vec::new(),
        }
    }

    pub fn learn(&mut self, fingerprint_key: &[u8], secret: &[u8]) -> String {
        let fingerprint = redaction_fingerprint(fingerprint_key, secret);
        self.known_values
            .push((fingerprint.clone(), secret.to_vec()));
        fingerprint
    }

    pub fn with_known_secret(fingerprint: impl Into<String>, secret: impl Into<Vec<u8>>) -> Self {
        Self {
            known_values: vec![(fingerprint.into(), secret.into())],
        }
    }

    pub fn redact_text(&self, text: &str) -> String {
        let mut redacted = text.to_string();
        for (fingerprint, secret) in &self.known_values {
            let marker = format!("[REDACTED:secret:{fingerprint}]");
            let mut representations = secret_representations(secret);
            representations.sort_by_key(|value| std::cmp::Reverse(value.len()));
            for representation in representations {
                redacted = redacted.replace(&representation, &marker);
            }
        }
        redact_structured_patterns(&redacted)
    }

    pub fn assert_no_known_secret(&self, text: &str) -> Result<(), RedactionLeakError> {
        if self.redact_text(text) == text {
            Ok(())
        } else {
            Err(RedactionLeakError)
        }
    }
}

impl Default for RedactionEngine {
    fn default() -> Self {
        Self::new()
    }
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, thiserror::Error)]
#[error("controlled output contains a known secret")]
pub struct RedactionLeakError;

pub fn redact_known_secret(value: &str, secret: &[u8], marker: &str) -> (String, bool) {
    let fingerprint = marker.to_string();
    let engine = RedactionEngine::with_known_secret(fingerprint, secret.to_vec());
    let redacted = engine.redact_text(value);
    let changed = redacted != value;
    (redacted, changed)
}

pub fn secret_representations(secret: &[u8]) -> Vec<String> {
    let mut values = vec![base64_encode(secret), b64url_no_padding(secret)];
    if let Ok(text) = std::str::from_utf8(secret) {
        values.push(text.to_string());
        values.push(percent_encode(text));
    }
    values.sort();
    values.dedup();
    values
        .into_iter()
        .filter(|value| !value.is_empty())
        .collect()
}

fn redact_structured_patterns(text: &str) -> String {
    let patterns = [
        (
            r"(?i)(Authorization:\s*Bearer\s+)[^\s,;]+",
            "$1[REDACTED:bearer]",
        ),
        (
            r#"(?i)(password=)(["']?)[^;&\s"']+(["']?)"#,
            "$1$2[REDACTED]$3",
        ),
        (
            r#"(?i)(token=)(["']?)[^;&\s"']+(["']?)"#,
            "$1$2[REDACTED]$3",
        ),
        (
            r#"(?i)(api[_-]?key=)(["']?)[^;&\s"']+(["']?)"#,
            "$1$2[REDACTED]$3",
        ),
        (
            r#"(?i)("?(?:password|token|api[_-]?key)"?\s*:\s*")[^"]+(")"#,
            "$1[REDACTED]$2",
        ),
    ];
    let mut redacted = text.to_string();
    for (pattern, replacement) in patterns {
        redacted = Regex::new(pattern)
            .expect("static redaction regex compiles")
            .replace_all(&redacted, replacement)
            .to_string();
    }
    redacted
}

fn base64_encode(value: &[u8]) -> String {
    use base64::Engine;
    base64::engine::general_purpose::STANDARD.encode(value)
}

fn percent_encode(value: &str) -> String {
    let mut encoded = String::new();
    for byte in value.as_bytes() {
        if byte.is_ascii_alphanumeric() || matches!(byte, b'-' | b'_' | b'.' | b'~') {
            encoded.push(*byte as char);
        } else {
            encoded.push_str(&format!("%{byte:02X}"));
        }
    }
    encoded
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn fingerprint_is_keyed_and_truncated() {
        let first = redaction_fingerprint(&[1u8; 32], b"secret");
        let second = redaction_fingerprint(&[2u8; 32], b"secret");
        assert_ne!(first, second);
        assert_eq!(first.len(), 24);
    }

    #[test]
    fn redacts_known_representations_and_structured_patterns() {
        let mut engine = RedactionEngine::new();
        engine.learn(&[1u8; 32], b"sk test");
        assert_eq!(engine.redact_text("token=sk test"), "token=[REDACTED]");
        assert_eq!(
            engine.redact_text("Authorization: Bearer abc123"),
            "Authorization: Bearer [REDACTED:bearer]"
        );
        assert!(engine
            .assert_no_known_secret("prefix sk test suffix")
            .is_err());
    }
}
